icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,536 @@
+    int32_t end = fUnicodeBlock->size();
+    while (start < end) {
+        int32_t mid = (start + end) / 2;
+        if (ch < fUnicodeBlock->elementAti(mid)) {
+            end = mid;
+        }
+        else {
+            start = mid + 1;
+        }
+    }
+    //Add '0' to fill up three digits
+    char buff[32];
+    itoa(start, buff);
+    UnicodeString result(buff);
+    if (result.length() < 3) {
+        int32_t preZeroCount = 3 - result.length();
+        for (int i = 0; i < preZeroCount; i++) {
+            result.insert(0, '0');
+        }
+    }
+    return result;
+}
+
+void CjMLBreakEngine::loadMLparams(UErrorCode& error) {
+    negativeSum = 0;
+    const char* tags[] = { "model", "unicode_blocks" };
+    ResourceBundle ja(U_ICUDATA_BRKITR, "ja", error);
+    UnicodeString key, value;
+    UChar32 delimeter = ':';
+    char charBuf[512];
+    charBuf[sizeof(charBuf) - 1] = 0;
+    int32_t iValue = 0;
+    if (U_SUCCESS(error)) {
+        ResourceBundle bundle = ja.get(tags[0], error);
+        while (U_SUCCESS(error) && bundle.hasNext()) {
+            UnicodeString temp = bundle.getNextString(error);
+            int32_t index = temp.lastIndexOf(delimeter);
+            key = temp.tempSubString(0, index);
+            value = temp.tempSubString(index + 1, temp.length() - index - 1);
+            //transform to integer
+            value.extract(0, value.length(), charBuf, sizeof(charBuf) - 1, 0);
+            iValue = atoi(charBuf);
+            negativeSum -= iValue;
+            fModel.puti(key, iValue, error);
+        }
+    }
+    if (U_SUCCESS(error)) {
+        ResourceBundle bundle = ja.get(tags[1], error);
+        while (U_SUCCESS(error) && bundle.hasNext()) {
+            UnicodeString value = bundle.getNextString(error);
+            //transform to integer
+            value.extract(0, value.length(), charBuf, sizeof(charBuf) - 1, 0);
+            fUnicodeBlock->addElement(atoi(charBuf), error);
+        }
+    }
+}
+
+ThreadPool::ThreadPool() { busy = 0; }
+ThreadPool::~ThreadPool(){}
+
+void ThreadPool::loop() {
+    while (true) {
+        std::unique_lock<std::mutex> latch(queue_mutex);
+        mutex_job.wait(latch, [this]() { return should_terminate || !jobs.empty(); });
+        if (!jobs.empty()) {
+            ++busy;
+
+            auto job = jobs.front();
+            jobs.pop();
+
+            // release lock. run async
+            latch.unlock();
+
+            // run function outside context
+            job();
+
+            latch.lock();
+            --busy;
+            mutex_finish.notify_one();
+        } else if (should_terminate) {
+            break;
+        }
+    }
+}
+
+void ThreadPool::start() {
+    int32_t threadCount = std::thread::hardware_concurrency();
+    int32_t size = poolSize < threadCount ? poolSize : threadCount;
+    threads.resize(size);
+    for (int32_t i = 0; i < size; i++) {
+        threads.push_back(std::thread(&ThreadPool::loop, this));
+    }
+}
```

Comment:
This... is something that ICU has never done before. ICU has its own wrappers for mutexes and once-init and atomic-increment functionality (internal [mutex.h](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/mutex.h) & [umutex.h](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/umutex.h)), but ICU library code has never spawned new threads before. Is this necessary?

@FrankYFTang @aheninger @pedberg-icu @richgillam @jefgen

Reply:
Hi Markus

To improve the performance, we use two approaches
1. Fine tune the ML model. (improve ~40%)
2. Apply parallel processing, ThreadPool. (improve 10~ 20%)
https://docs.google.com/document/d/1mo9AZHfvPvd-oKgFh9qN6DgUe1I4pOW8ibmK44b6T0w/edit#heading=h.s2giaj6oewfc

That's why I added std::thread.
Since it can improve 10~20%, please confirm if I can use it in ICU.
mutex, queue and the change in dependencies.txt all result from this thread design.

Thanks

Reply:
Done. Remove it.